chain/chain/src/runtime/mod.rs

```diff
@@ -407,18 +406,42 @@ impl NightshadeRuntime {
             part_id,
             trie_with_state,
         );
-        let state_part = borsh::to_vec(&match trie_nodes {
+        let partial_state = match trie_nodes {
             Ok(partial_state) => partial_state,
             Err(err) => {
                 error!(target: "runtime", ?err, part_id.idx, part_id.total, %prev_hash, %state_root, %shard_id, "Can't get trie nodes for state part");
                 return Err(err.into());
             }
-        })
-            .expect("serializer should not fail");
-
+        };
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        let state_part = StatePart::from_partial_state(partial_state, protocol_version);
         Ok(state_part)
     }
 
+    fn validate_state_part_impl(
+        &self,
+        state_root: &StateRoot,
+        part_id: PartId,
+        part: &StatePart,
+    ) -> bool {
```

Comment:
nit: It's usually nicer to return an enum e.g. StatePartValidationResult { Valid, Invalid }

Reply:
We return `bool` in many places like `validate_state_root_node`, `validate_state_part` and the result is checked in many tests. To make this PR more readable, I created an [issue](https://github.com/near/nearcore/issues/14047) to deal with that later, e.g. next code quality hackathon.